crates/dbsp/src/operator/dynamic/saturate.rs

```diff
@@ -0,0 +1,281 @@
+        //  stream ┌──────────┐   │   ┌─────┐  delayed trace  ┌────────┐ tuples    ┌──────────┐
+        // ───────►│accumulate├───┴──►│trace├────────────────►│Saturate├──────────►│accumulate│
+        //         └────┬─────┘       └─────┘                 └────────┘           └─────┬────┘
+        //              │                                                                │
+        //              │                                                                ▼
+        //              │                                                            ┌──────┐
+        //              └───────────────────────────────────────────────────────────►│  +   ├──────►
+        //                                                                           └──────┘
+        // ```
+
+        self.circuit().region("saturate", || {
+            let stream = self.dyn_shard(&factories.batch_factories);
+
+            let delayed_trace = stream
+                .dyn_accumulate_trace(&factories.trace_factories, &factories.batch_factories)
+                .accumulate_delay_trace();
+
+            let ghost = self.circuit().add_binary_operator(
+                StreamingBinaryWrapper::new(Saturate::new(&factories.batch_factories)),
+                &stream.dyn_accumulate(&factories.batch_factories),
+                &delayed_trace,
+            );
+
+            ghost.mark_sharded();
+
+            let output_factories = factories.batch_factories.clone();
+
+            // Plus
+            let result = stream.circuit().add_binary_operator(
+                AccumulateApply2::new(
+                    move |stream, saturation| {
+                        SpineSnapshot::concat(output_factories.clone(), vec![&stream, &saturation])
+                    },
+                    Location::caller(),
+                ),
+                &stream.dyn_accumulate(&factories.batch_factories),
+                &ghost.dyn_accumulate(&factories.batch_factories),
+            );
+
+            result.mark_sharded();
+            result
+        })
+    }
+}
+
+/// This operator computes the ghost tuples. Concatenate its output with the original stream
+/// to obtain the complete saturated stream.
+struct Saturate<K, V>
+where
+    K: DataTrait + ?Sized,
+    V: DataTrait + ?Sized,
+{
+    factories: OrdIndexedZSetFactories<K, V>,
+    input_batch_stats: BatchSizeStats,
+    output_batch_stats: RefCell<BatchSizeStats>,
+    phantom: PhantomData<fn(&K, &V)>,
+}
+
+impl<K, V> Saturate<K, V>
+where
+    K: DataTrait + ?Sized,
+    V: DataTrait + ?Sized,
+{
+    pub fn new(factories: &OrdIndexedZSetFactories<K, V>) -> Self {
+        Self {
+            factories: factories.clone(),
+            input_batch_stats: BatchSizeStats::new(),
+            output_batch_stats: RefCell::new(BatchSizeStats::new()),
+            phantom: PhantomData,
+        }
+    }
+
+    /// True if there's at least one value with a non-zero weight in the combined stream.
```

Comment:
the implementation seems to check that the stream intersection is not empty
is that right?

Reply:
it checks union, not intersection.